pre_commit/languages/pygrep.py

```diff
@@ -17,31 +17,47 @@
 install_environment = helpers.no_install
 
 
-def _process_filename_by_line(pattern: Pattern[bytes], filename: str) -> int:
+def _process_filename_by_line(
+    pattern: Pattern[bytes],
+    filename: str,
+    negate: bool,
+) -> int:
     retv = 0
     with open(filename, 'rb') as f:
         for line_no, line in enumerate(f, start=1):
             if pattern.search(line):
                 retv = 1
-                output.write(f'{filename}:{line_no}:')
-                output.write_line_b(line.rstrip(b'\r\n'))
+                if not negate:
+                    output.write(f'{filename}:{line_no}:')
+                    output.write_line_b(line.rstrip(b'\r\n'))
+    if negate and not retv:
+        output.write_line(filename)
     return retv
 
 
-def _process_filename_at_once(pattern: Pattern[bytes], filename: str) -> int:
+def _process_filename_at_once(
+    pattern: Pattern[bytes],
+    filename: str,
+    negate: bool,
+) -> int:
     retv = 0
     with open(filename, 'rb') as f:
         contents = f.read()
         match = pattern.search(contents)
         if match:
             retv = 1
-            line_no = contents[:match.start()].count(b'\n')
-            output.write(f'{filename}:{line_no + 1}:')
+            if not negate:
+                line_no = contents[:match.start()].count(b'\n')
+                output.write(f'{filename}:{line_no + 1}:')
 
-            matched_lines = match[0].split(b'\n')
-            matched_lines[0] = contents.split(b'\n')[line_no]
+                matched_lines = match[0].split(b'\n')
+                matched_lines[0] = contents.split(b'\n')[line_no]
+
+                output.write_line_b(b'\n'.join(matched_lines))
+        else:
+            if negate:
+                output.write_line(filename)
 
-            output.write_line_b(b'\n'.join(matched_lines))
     return retv
```

Comment:
these functions now both are pretty hard to understand and have a [boolean trap](https://ariya.io/2011/08/hall-of-api-shame-boolean-trap) -- I'd suggest refactoring them into 4 separate easier to understand functions

Reply:
Thanks for your review, and for another really good link!

Sure, I've made another attempt at this